tidb-latency-formula.md

```diff
@@ -0,0 +1,531 @@
+# TiDB Latency Formula
+
+This document breaks down the latency into metrics, for those who have interests in observation of the metrics, it's a guide of TiDB's critical path diagnosis.
+
+Generally, OLTP workload can be divided into read and write queries, they share some critical code.
+This document will break down the latency and then analyze it from user's perspective. It's better to read [Performance Analysis and Tuning](/performance-tuning-methods.md) before this document.
+Note when breaking down latency through metrics in this document, we are calculating the average duration or latency instead of some specific slow queries.
+Many metrics are collected as histogram, which is a distribution of the duration or latency, when calculating the average latency, you need to use the sum and count counter.
+
+```
+avg = ${metric_name}_sum / ${metric_name}_count
+```
+
+Metrics in this document can be read directly from prometheus of TiDB.
+
+## General SQL Layer
+
+This part of latency is on the top level of TiDB and shared by any queries.
+
+```
+e2e duration =
+    tidb_server_get_token_duration_seconds +
+    tidb_session_parse_duration_seconds +
+    tidb_session_compile_duration_seconds +
+    tidb_session_execute_duration_seconds{type="general"}
+```
+
+`tidb_server_get_token_duration_seconds` should be small enough to be ignored, usually less than 1 microsecond.
+
+`tidb_session_parse_duration_seconds` records the duration pf parsing sql query text to AST, which can be skipped by `prepare/execute` statements.
```

Comment:
We can link it to dev-guide-optimize-sql-best-practices.md#use-prepare like the compile below.